src/USER-CONP/compute_conp_vector.cpp

```diff
@@ -305,94 +270,49 @@ void ComputeConpVector::pair_contribution() {
 
   int *mask = atom->mask;
-  int nlocal = atom->nlocal;
-  int nall = atom->nghost + atom->nlocal;
-  int nprocs = comm->nprocs;
+  int const nlocal = atom->nlocal;
+  int const nprocs = comm->nprocs;
   tagint *tag = atom->tag;
-  int igroupnum_local = 0, jgroupnum_local = 0;
 
+  std::vector<int> taglist_local;
   for (int i = 0; i < nlocal; i++) {
-    if (mask[i] & groupbit && mask[i] & jgroupbit)
-      error->all(FLERR, "Same atom in both groups in compute coul/matrix");
-    else if (mask[i] & groupbit)
-      igroupnum_local++;
-    else if (mask[i] & jgroupbit)
-      jgroupnum_local++;
+    if (mask[i] & groupbit) taglist_local.push_back(tag[i]);
   }
+  int igroupnum_local = taglist_local.size();
 
   std::vector<int> idispls(nprocs);
-  std::vector<int> jdispls(nprocs);
   std::vector<int> igroupnum_list(nprocs);
-  std::vector<int> jgroupnum_list(nprocs);
-
   MPI_Allgather(&igroupnum_local, 1, MPI_INT, &igroupnum_list.front(), 1,
                 MPI_INT, world);
-  MPI_Allgather(&jgroupnum_local, 1, MPI_INT, &jgroupnum_list.front(), 1,
-                MPI_INT, world);
-
-  idispls[0] = jdispls[0] = 0;
+  idispls[0] = 0;
   for (int i = 1; i < nprocs; i++) {
     idispls[i] = idispls[i - 1] + igroupnum_list[i - 1];
-    jdispls[i] = jdispls[i - 1] + jgroupnum_list[i - 1];
   }
 
-  std::vector<int> itaglist_local(igroupnum_local);
-  std::vector<int> jtaglist_local(jgroupnum_local);
-
-  igroupnum_local = jgroupnum_local = 0;
-  for (int i = 0; i < nlocal; i++) {
-    if (mask[i] & groupbit) {
-      itaglist_local[igroupnum_local] = tag[i];
-      igroupnum_local++;
-    } else if (mask[i] & jgroupbit) {
-      jtaglist_local[jgroupnum_local] = tag[i];
-      jgroupnum_local++;
-    }
-  }
-
-  std::vector<int> itaglist(igroupnum);
-  std::vector<int> jtaglist(jgroupnum);
-
-  MPI_Allgatherv(&itaglist_local.front(), igroupnum_local, MPI_LMP_TAGINT,
-                 &itaglist.front(), &igroupnum_list.front(), &idispls.front(),
-                 MPI_LMP_TAGINT, world);
-  MPI_Allgatherv(&jtaglist_local.front(), jgroupnum_local, MPI_LMP_TAGINT,
-                 &jtaglist.front(), &jgroupnum_list.front(), &jdispls.front(),
+  taglist = std::vector<int>(ngroup);
+  MPI_Allgatherv(&taglist_local.front(), igroupnum_local, MPI_LMP_TAGINT,
+                 &taglist.front(), &igroupnum_list.front(), &idispls.front(),
                  MPI_LMP_TAGINT, world);
+  std::sort(taglist.begin(), taglist.end());
+}
 
-  // sort individual group taglists, first igroup than jgroup
-
-  std::sort(itaglist.begin(), itaglist.end());
-  std::sort(jtaglist.begin(), jtaglist.end());
-
-  // if local+ghost matrix assignment already created, recreate
-
-  if (assigned) {
-    memory->destroy(mpos);
-  }
-  memory->create(mpos, nall, "coul/vector:mpos");
+/* ---------------------------------------------------------------------- */
 
+void ComputeConpVector::update_mpos() {
+  int const nall = atom->nlocal + atom->nghost;
```

Comment:
Für nall kannst du nmax nehmen. Das ist in atom->nmax und umfasst genau die Atome aus nlocal+nghost. Ist etwas sauberer, wenn du das so machst.